### PR TITLE
chainparams: Add achow101 DNS seeder

### PR DESCRIPTION
**Pull request by achow101**
I wrote a [DNS seeder](https://github.com/achow101/dnsseedrs) and have been running it for the past 2 months now. I believe it is ready/good enough to be used as an additional DNS seeder for all of our supported public networks.

### CODE REVIEW

File: src/kernel/chainparams.cpp [hunk 1]
```diff
@@ -246,6 +247,8 @@ class CTestNetParams : public CChainParams {
         vSeeds.emplace_back("seed.tbtc.petertodd.net.");
         vSeeds.emplace_back("seed.testnet.bitcoin.sprovoost.nl.");
         vSeeds.emplace_back("testnet-seed.bluematt.me."); // Just a static list of stable node(s), only supports x9
+        vSeeds.emplace_back("dnsseed.testnet.bitcoin.achow101.com"); // Ava Chow, only supports x1, x5, x9, x49, x809, x849, xd, x400, x404, x408, x448, xc08, xc48, x40c
```
Comment (laanwj) [inline src/kernel/chainparams.cpp:250]: Looks like it's missing "." at the end (all the other seeder DNS names have this)--same for the testnet one

Comment (achow101) [reply src/kernel/chainparams.cpp]: Done

**Comment by laanwj** [comment]
Concept ACK
```
Flags: x9
Status DNS name                                 Totals     IPv4                  IPv6
                                                nconn/n    nconn/n           TTL nconn/n           TTL
* mainnet
OK     seed.bitcoin.sipa.be.                    25/25      25/25            2640 0/0
OK     dnsseed.bluematt.me.                     29/31      20/21              60 9/10               60
OK     dnsseed.bitcoin.dashjr-list-of-p2p-nodes 22/22      22/22            2644 0/0
ERR    seed.bitcoinstats.com.                   0/0        SERVFAIL              SERVFAIL
ERR    seed.bitcoin.jonasschnelli.ch.           0/0        SERVFAIL              SERVFAIL
OK     seed.btc.petertodd.net.                  11/36      7/23             2705 4/13             3600
OK     seed.bitcoin.sprovoost.nl.               22/36      15/23            2705 7/13             3600
OK     dnsseed.emzy.de.                         35/39      21/25            2764 14/14            3600
OK     seed.bitcoin.wiz.biz.                    23/31      16/21              60 7/10               60
OK     dnsseed.mainnet.bitcoin.achow101.com.    38/40      20/20              30 18/20              30

* testnet
NONE   testnet-seed.bitcoin.jonasschnelli.ch.   0/0        0/0                   0/0
OK     seed.tbtc.petertodd.net.                 15/36      9/23             2841 6/13             3600
NONE   testnet-seed.bluematt.me.                0/0        0/0                   0/0
NONE   seed.testnet.bitcoin.sprovoost.nl.       0/0        0/0                   0/0
OK     dnsseed.testnet.bitcoin.achow101.com.    31/40      15/20              30 16/20              29

* signet
NONE   seed.signet.bitcoin.sprovoost.nl.        0/0        0/0                   0/0
OK     dnsseed.signet.bitcoin.achow101.com.     30/40      15/20              30 15/20              30
```

- Returns a good number of results (20 per query), of which a large fraction tends to be connectable within one second.
- A TTL of 30 seconds is low compared to the others. FWIW, in [dnsseed-policy.md](https://github.com/bitcoin/bitcoin/blob/master/doc/dnsseed-policy.md) a minimum of 60 seconds is mentioned.

**Comment by willcl-ark** [comment]
Concept ACK

Some of our DNS seeds are currently not performing [well](https://github.com/bitcoin/bitcoin/issues/29911), so adding a new/more reliable one seems logical to me.

I have also been running this seeder myself for some time (at seed.bitcoin.fish.foo) and the program seems to work well from the operator side too, not requiring any intervention in the few weeks i've been running it.

I ran a different test to @laanwj on mainnet IPV4 only, and did find seeds generally returning results:

<details>
<summary>Details</summary>

```
Results:

seed.bitcoin.sipa.be.
  x1         responses: 25
  x5         responses: 25
  x9         responses: 25
  xd         responses: 25

dnsseed.bluematt.me.
  x9         responses: 21

dnsseed.bitcoin.dashjr-list-of-p2p-nodes.us.
  (no flag)  responses: 22
  x9         responses: 22

seed.bitcoinstats.com.
  x1         responses: 17
  x2         responses: 17
  x3         responses: 17
  x4         responses: 17
  x5         responses: 17
  x6         responses: 17
  x7         responses: 17
  x8         responses: 17
  x9         responses: 17
  xa         responses: 17
  xb         responses: 17
  xc         responses: 17
  xd         responses: 17
  xe         responses: 17
  xf         responses: 17

seed.bitcoin.jonasschnelli.ch.
  x1         responses: 23
  x5         responses: 23
  x9         responses: 23
  xd         responses: 23

seed.btc.petertodd.net.
  x1         responses: 23
  x5         responses: 23
  x9         responses: 23
  xd         responses: 23

seed.bitcoin.sprovoost.nl.
  (no flag)  responses: 23
  x9         responses: 23

dnsseed.emzy.de.
  (no flag)  responses: 26
  x9         responses: 25

seed.bitcoin.wiz.biz.
  (no flag)  responses: 21
  x9         responses: 21

dnsseed.mainnet.bitcoin.achow101.com.
  x1         responses: 20
  x5         responses: 20
  x9         responses: 20
  x49        responses: 20
  x809       responses: 20
  x849       responses: 20
  xd         responses: 20
  x400       responses: 20
  x404       responses: 20
  x408       responses: 20
  x448       responses: 20
  xc08       responses: 20
  xc48       responses: 20
  x40c       responses: 20
```

</details>

I will try to expand my test script soon to include testnet and also to query whether the flags are accurate for returned results so I can verify both this new seed, and existing seeds.

**Comment by achow101** [comment]
> A TTL of 30 seconds is low compared to the others. FWIW, in [dnsseed-policy.md](https://github.com/bitcoin/bitcoin/blob/master/doc/dnsseed-policy.md) a minimum of 60 seconds is mentioned.

I've changed it to 60 seconds.

**Comment by mzumsande** [comment]
@virtu FYI, would it be easily possible to run some of the metrics of https://21.ninja/dns-seeds/ for this new seeder?

**Comment by ariard** [comment]
Concept ACK.

By the way, it would be great if mainnet DNS seeders are considering to sign by default the peers records. This could be amply checked in `ThreadDNSAddressSeed()` or in semi-automatic fashion in the logs. ECDSA secp256k1 isn’t supported by default in DNS, though you have ed25519 or ECDSA P256 which are widely supported by key management softwares.

**Comment by laanwj** [comment]
> By the way, it would be great if mainnet DNS seeders are considering to sign by default the peers records. This could be amply checked in ThreadDNSAddressSeed()

There are some that do (#19714), but as far as i know, there is no cross-platform API for checking DNSSEC status from user code. i've unlocked that issue for discussion.

**Comment by virtu** [comment]
> @virtu FYI, would it be easily possible to run some of the metrics of https://21.ninja/dns-seeds/ for this new seeder?

@mzumsande, the seed is now being monitored on [dev.21.ninja](dev.21.ninja/dns-seeds/).

There may be some graph artifacts until a second data point becomes available. But so far data looks good: 40 advertised addresses (half of them ipv4, the other ipv6), and 35 of them reachable.

**Comment by 1440000bytes** [comment]
Why does the seeder consider 'default port' for good nodes?

**Comment by achow101** [comment]
> Why does the seeder consider 'default port' for good nodes?

DNS cannot provide port numbers, but a port must be known when connecting to a node. So we assume the default port, and because of that assumption, DNS seeders need to return nodes that are listening on the default port.

**Comment by Sjors** [comment]
Concept ACK

There's discussion in #29911 about whether we should mention the specific feature bits here.

I tested that the mainnet seed result returns both IPv4 and IPv6 records and tried to connect to a random result. I didn't do any fancier analysis.

**Comment by 1440000bytes** [comment]
Concept ACK on adding another DNS seeder

> > Why does the seeder consider 'default port' for good nodes?
> 
> DNS cannot provide port numbers, but a port must be known when connecting to a node. So we assume the default port, and because of that assumption, DNS seeders need to return nodes that are listening on the default port.

TXT records could work but that will require lot of other changes (out of scope)

**Comment by laanwj** [comment]
> TXT records could work but that will require lot of other changes (out of scope)

Right-DNS can serve arbitrary information, but it would complicate things in the client: the cross-platform libc resolver, `getaddrinfo`, can only resolve addresses. So for that'd we'd need to add some dependency library for DNS resolving. The question is whether it's worth it.

Also, IIRC caching DNS servers don't always cache TXT records (*because* they can store arbitrary data); the caching and the privacy that comes with it, is the advantage of using DNS in the first place, over simply using bitcoin protocol for seeding.

Mind that the DNS seeds are only an entry point to the network. The gossip network itself can handle alternative ports fine, so from that point on, nodes with other ports can be discovered by a node. In the main threat scenario that would make this entry point useless, a hypothetical future where port 8333 would be blocked by ISPs, it's extremely likely that the DNS seeds would also be blocked entirely as they're easy to enumerate.

**Comment by achow101** [comment]
I've implemented DNSSEC

**Comment by laanwj** [comment]
> I've implemented DNSSEC

That's neat!

i think it's still missing some part, resolving through Google's DNS (which has more verbose error messages than my ISP) gives:
```sh
$ dig x9.dnsseed.signet.bitcoin.achow101.com. @1.1.1.1
⋮
;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 1232
; EDE: 10 (RRSIGs Missing): (failed to verify signatures for x9.dnsseed.signet.bitcoin.achow101.com. opt-out proof)
⋮
```

**Comment by achow101** [comment]
> i think it's still missing some part, resolving through Google's DNS (which has more verbose error messages than my ISP) gives:

Should be fixed now

**Review by laanwj — APPROVED** [review]
All good now!

ACK ee218aa9a9eaac53030c31b099b4afe354197ba7

**Review by 1440000bytes — APPROVED** [review]
ACK https://github.com/bitcoin/bitcoin/pull/30007/commits/ee218aa9a9eaac53030c31b099b4afe354197ba7

**Comment by luke-jr** [comment]
@achow101 Are you sure you want to put this on a common domain with other things?

**Comment by kristapsk** [comment]
Concept ACK on using different software for various DNS seeders. Need to do more review / testing on this one.

**Review by willcl-ark — APPROVED** [review]
ACK ee218aa9a9eaac53030c31b099b4afe354197ba7

I checked that these seeds return diverse and (mainly) active node addresses (although not in a scripted way, but by `addnode`-ing a random selection of each).

I also checked DNSSEC setup using Verisign's DNSSEC [debugger](https://dnssec-debugger.verisignlabs.com) tool for each of them, which they all passed.

**Comment by achow101** [comment]
Anything left to do here?

***

> Are you sure you want to put this on a common domain with other things?

I don't think this will be an issue.

**Comment by achow101** [comment]
> > Are you sure you want to put this on a common domain with other things?
> 
> I don't think this will be an issue.

Actually on second thought I will switch this to a different domain.

**Comment by achow101** [comment]
Updated root domain to `achownodes.xyz`

**Review by 1440000bytes — APPROVED** [review]
~~reACK https://github.com/bitcoin/bitcoin/pull/30007/commits/2721d64989c2b2114890586b7efd01ab4b062ca6~~

**Review by mzumsande — COMMENTED** [review]
ACK 2721d64989c2b2114890586b7efd01ab4b062ca6

**Review by willcl-ark — APPROVED** [review]
reACK 2721d64989c2b2114890586b7efd01ab4b062ca6

Retested that the seed on the new domain for each chain is returning good addresses.

**Comment by 1440000bytes** [comment]
I was not sure about .xyz domain for a bitcoin DNS seed based on things that I have read. Still ACKed it because I have been using .xyz domain for joinstr and had no issues. After recent fedimint incident I think `achownodes.xyz` will also get suspended sooner or later. So retracted my [ACK](https://github.com/bitcoin/bitcoin/pull/30007#pullrequestreview-2120313513).

DNS seed domains resolve to lot of IP addresses which may host some website. Its possible that someone will report it or it will get red flagged by automated systems.

<details>
  <summary>Screenshots</summary>
  <p>
    Image 1
    Image 2
  </p>
</details>

**Comment by achow101** [comment]
> After recent fedimint incident I think `achownodes.xyz` will also get suspended sooner or later.

That is a risk of using anything DNS as the registry operator can seize them. But I think it's preferable to have more DNS seeders so that we are not overly reliant on a few of them. Furthermore, DNS seeders are not critical infrastructure - one being shut down does not really impact the network that much. They're also easily portable and changing domains is not difficult.

> DNS seed domains resolve to lot of IP addresses which may host some website. Its possible that someone will report it or it will get red flagged by automated systems.

The actual sites on this domain have HSTS enabled with subdomain inclusion, although you have to visit those sites for the browser to know about that. I've also submitted it to the HSTS preload list. Once a browser knows it has HSTS enabled (whether by visiting the actual sites or through the preload list), the random sites people have on their node IPs should stop loading as they will not have a valid certificate for the domain.

**Comment by laanwj** [comment]
ACK 2721d64989c2b2114890586b7efd01ab4b062ca6
Adding a DNS seed adds redundancy, so does spreading them over as many top-level domains as possible, so i don't think "this domain might get blocked" is ever an argument against this.

**Comment by Sjors** [comment]
> The actual sites on this domain have HSTS enabled with subdomain inclusion, although you have to visit those sites for the browser to know about that. I've also submitted it to the HSTS preload list.

Oh that's a good idea...

**Pull request merged**